libs/langchain/langchain/retrievers/google_cloud_documentai_warehouse.py

```diff
@@ -0,0 +1,126 @@
+        SearchDocumentsPager,
+    )
+
+
+class GoogleDocumentAIWarehouseRetriever(BaseRetriever):
+    """A retriever based on Document AI Warehouse.
+
+    Documents should be created and documents should be uploaded
+        in a separate flow, and this retriever uses only Document AI
+        schema_id provided to search for revelant documents.
+
+    More info: https://cloud.google.com/document-ai-warehouse.
+    """
+
+    location: str = "us"
+    "GCP location where DocAI Warehouse is placed."
+    project_id: str
+    "GCP project number, should contain digits only."
+    schema_id: Optional[str] = None
+    "DocAI Warehouse schema to queary against. If nothing is provided, all documents "
+    "in the project will be searched."
+    qa_size_limit: int = 5
+    "The limit on the number of documents returned."
+    client: "DocumentServiceClient" = None  #: :meta private:
+
+    @root_validator()
+    def validate_environment(cls, values: Dict) -> Dict:
+        """Validates the environment."""
+        try:  # noqa: F401
+            from google.cloud.contentwarehouse_v1 import (
+                DocumentServiceClient,
+            )
+        except ImportError as exc:
+            raise ImportError(
+                "google.cloud.contentwarehouse is not installed."
+                "Please install it with pip install google-cloud-contentwarehouse"
+            ) from exc
+
+        values["project_id"] = get_from_dict_or_env(values, "project_id", "PROJECT_ID")
+        values["client"] = DocumentServiceClient()
+        return values
+
+    @property
+    def _parent(self) -> str:
+        return f"projects/{self.project_id}/locations/{self.location}"
+
+    @property
+    def _schemas(self) -> List[str]:
+        if self.schema_id:
+            return [f"{self._parent}/documentSchemas/{self.schema_id}"]
+        return []
+
+    def _prepare_request_metadata(self, user_ldap: str) -> "RequestMetadata":
+        from google.cloud.contentwarehouse_v1 import RequestMetadata, UserInfo
+
+        user_info = UserInfo(id=f"user:{user_ldap}")
+        return RequestMetadata(user_info=user_info)
+
+    def _get_relevant_documents(
+        self, query: str, *, run_manager: CallbackManagerForRetrieverRun, **kwargs: Any
+    ) -> List[Document]:
+        request = self._prepare_search_request(query, **kwargs)
+        response = self.client.search_documents(request=request)
+        return self._parse_search_response(response=response)
+
+    def _prepare_search_request(
+        self, query: str, **kwargs: Any
+    ) -> "SearchDocumentsRequest":
+        from google.cloud.contentwarehouse_v1 import (
+            DocumentQuery,
+            SearchDocumentsRequest,
+        )
+
+        try:
+            user_ldap = kwargs["user_ldap"]
+        except KeyError:
+            raise ValueError("Argument user_ldap should be provided!")
+
+        request_metadata = self._prepare_request_metadata(user_ldap=user_ldap)
```

Comment:
Does this really need to be a separate function? It's only a couple of lines and only used here.

Reply:
It helps with debugging.